Split main bundle

ecxyzzy opened the pull request:
## Summary
Reduce the size of the main bundle by chunking static assets, if possible.
* [x] Logo images
* [x] Building catalogue
* [x] Department list

## Test Plan
1. Navigate to the deployed production version and ensure that performance has improved.

## Issues
Closes #403.

<!-- [Optional]
## Future Followup  
-->

ecxyzzy committed:
lazy load logo svgs

ecxyzzy committed:
lazy load the entire UCIMap component

certified legacy tech stack moment

ecxyzzy committed:
prettify fallbacks

ecxyzzy committed:
remove suspense/lazy wrapper components

ecxyzzy committed:
consolidate styling

ecxyzzy committed:
fix styling issues

ecxyzzy commented on src/components/RightPane/CoursePane/SearchForm/SearchForm.js:
```diff
@@ -64,16 +75,27 @@ const SearchForm = (props) => {
-                            }}
-                            onReset={resetFormValues}
-                        />
+                        <Suspense
```
Also implemented lazy loading of legacy search in this commit, derp :upside_down_face:

ChaseC99 replied:
Why does this need a suspense? Are there any network calls?

It does look a bit strange having the course search loading gif displayed here temporarily

ecxyzzy replied:
I believe it's good practice to wrap lazy-loaded components in Suspenses, but it might also work without (haven't tested it).

I just decided to use the course search loading gif as a placeholder for the fallback, but we could just leave it blank as well.

ChaseC99 replied:
Oh yeah no you should really only lazy load something if there is a blocker in the component (e.g. an awaited network call). So a suspense should not be used here.

ecxyzzy replied:
Oh I remember why I wanted to lazy load legacy search now. It's because it uses the department search bar, which in turn includes the static department list, which was one of the stretch goals for #403.

Removing the suspense breaks things, but even an empty div in the fallback works, and should be less abrupt than the gif.
![image](https://user-images.githubusercontent.com/89349085/179048599-edb5dccf-a263-41b4-ab71-d0df82679350.png)

EricPedley replied:
With this change, bundle looks like this, with the split chunk for the legacy search highlighted:
![screenshot](https://user-images.githubusercontent.com/48658337/179343843-1c9f5b06-7e42-4043-8357-9edf1efc453d.png)
Without lazy loading legacy search:
![screenshot](https://user-images.githubusercontent.com/48658337/179343809-9960063f-aaed-4cae-bd69-4174aefb0435.png)
So we're just shaving off ~7 kb from the initial load, which is less than 1%
![screenshot](https://user-images.githubusercontent.com/48658337/179343888-d3a2fc2a-d188-405e-a047-a0ff2921dfb1.png)
I'm guessing the largest performance gain we're seeing from this PR here is lazy loading the UCIMap component and the other stuff is just marginal gains. I'll test that and reply in another comment.

EricPedley replied:
This is our bundle without lazy loading the map.
![image](https://user-images.githubusercontent.com/48658337/179344112-f3aaeb26-05a2-4c1c-ac48-753315616c03.png)
Looks like the bundle is roughly 60 kb bigger without the lazy loading, so I think this is where most of the performance is coming from.

EricPedley replied:
Also, reading the earlier conversation it seems like Chase really meant to ask, "why is this being lazy-loaded", because the response Eddy provided was basically that it needs a suspense is _because_ it is being lazy loaded, but I think that got cleared up. For extra clarification, the Suspense is needed to specify content to display while the lazy-loaded component is being loaded (in my understanding).

EricPedley replied:
I think we should just remove the lazy loading for the legacy search, given the negligible performance boost now that I actually did some research. Having our codebase remain as simple as possible is a bigger priority than a 1% performance boost. This probably means removing the lazy loading on the images too, but keeping it on the map since that does provide a significant boost.

EricPedley replied:
Now that I'm realizing we should prioratize based on file sizes, here are some other candidates for lazy loading and their uncompressed sizes:
- RightPane/CoursePane/static folder (the 'no results' images) - 70 kb
- SectionTable folder - 70 kb
- RightPane\CoursePane\SearchForm\Gifs - 2 MB
  - Maybe this gets relatively smaller on the bundle because image compression is better than source code compression, but I think this may be a major reason why our bundle is large. For reference, the entire source folder is 2.5 MB, so lazy loading this gets us potentially 80% size reduction at best.

ChaseC99 commented on src/components/AppBar/CustomAppBar.js:
```diff
@@ -44,7 +52,9 @@ const CustomAppBar = (props) => {
-                {isMobileScreen ? <MobileLogo height={32} /> : <Logo height={32} />}
+                <Suspense fallback={<div className={classes.fallback}>Loading...</div>}>
+                    {isMobileScreen ? <MobileLogo height={32} /> : <Logo height={32} />}
+                </Suspense>
```
The jump here is a bit jarring... 
The width of "Loading..." causes the buttons to compress and just looks a little strange in the bar.

![appbar_load](https://user-images.githubusercontent.com/29494270/178864702-9da26929-7620-4db9-b3d7-b7fe090bd00e.gif)

The logo svg is pretty small, do we even need to lazy load it?

ecxyzzy replied:
Ooh yeah that's something that needs fixing for sure, I didn't even notice it since my internet is so fast :sweat_smile: Something about living in darkness by going faster than light, et cetera.

As for lazy loading logos: that was one of the goals Eric mentioned in #403, so I'll defer to him on that.

ChaseC99 replied:
> I didn't even notice it since my internet is so fast 😅  

In your chrome dev tools, you can go to the network tab and throttle your internet speed to slow things down
![throttle](https://user-images.githubusercontent.com/29494270/179035840-628e3f1e-f9c7-462a-abca-fd6bf139d929.gif)

EricPedley replied:
> As for lazy loading logos: that was one of the goals Eric mentioned in #403, so I'll defer to him on that.

I put it on the issue because it looked like a good place to start. If it's going to be difficult to fix that rendering problem then we should remove the lazy loading, because it's such a small asset it only contributes a few kilobytes to the bundle size.

EricPedley replied:
After thinking about it more, we should definitely not lazy load these images. Doing so adds extra complexity for almost no benefit.

ChaseC99 commented on src/components/AppBar/Logo.js:
```diff
@@ -0,0 +1,3 @@
+import { ReactComponent as Logo } from './logo.svg';
+
+export default Logo;
```
If we decide not to lazy load this, we can get rid of Logo.js and MobileLogo.js

EricPedley commented on src/components/RightPane/Map/LazyUCIMap.js:
```diff
@@ -0,0 +1,12 @@
+import React, { Suspense } from 'react';
+
+const LazyUCIMap = () => {
```
What's the thought behind making this its own component? In my opinion, putting this inline in `RightPaneRoot` makes more sense since it's only a few lines and not reused anywhere.

ecxyzzy replied:
This should no longer be in the PR as of daae514, as it's been consolidated into `RightPaneRoot`. I think GitHub might have given you an outdated version of the source tree to review, since my requesting of your review predated that commit.

EricPedley reviewed (changes requested):
We should roll back the lazy loading on the legacy search and logo images, and look into lazy loading these components (same list as in my comment on the LegacySearch thread):
- RightPane/CoursePane/static folder (the 'no results' images) - 70 kb
- SectionTable folder - 70 kb
- RightPane\CoursePane\SearchForm\Gifs - 2 MB
  - Maybe this gets relatively smaller on the bundle because image compression is better than source code compression, but I think this may be a major reason why our bundle is large. For reference, the entire source folder is 2.5 MB, so lazy loading this gets us potentially 80% size reduction at best.
 
The other action item is replying to me about why LazyUCIMap needs to be its own component. Other than that you can skip my comments unless you're interested in seeing my thought process and rationale behind decisions.

ecxyzzy committed:
rollback lazy loading of LegacySearch

ecxyzzy committed:
Revert "lazy load logo svgs"

This reverts commit f96c8ba59ecad6a0afa92cba97831a89468d3819.

ecxyzzy committed:
lazy load section table with wrapper

Rationale: Since the SectionTable component is used by multiple
different components, unlike the old LazyUCIMap component, it should be
its own component to avoid repetition of boilerplate code.

EricPedley reviewed (approved):
![image](https://user-images.githubusercontent.com/48658337/179369294-9c7bc485-ee4f-4e86-bf1f-2ce7a481f7f7.png)
Very big reduction in bundle size from this one. I think we can call this PR good to go.
For reference, here was the original bundle size (before any lazy loading):
![image](https://user-images.githubusercontent.com/48658337/179369371-1ba47672-8939-4c39-a67c-0b22693fbdd5.png)
So it's been reduced by roughly 30%

EricPedley committed:
fix warning about no key for component in list
